File: src/test/java/org/zenbot/szolnok/timetable/batch/listener/RemoveBusRoutesExecutionListenerTest.java
package org.zenbot.szolnok.timetable.batch.listener;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.springframework.batch.core.JobExecution;
import org.zenbot.szolnok.timetable.configuration.properties.TimetableResourceProperties;
import org.zenbot.szolnok.timetable.dao.BusRepository;
import org.zenbot.szolnok.timetable.dao.BusStopRepository;
import org.zenbot.szolnok.timetable.domain.Bus;
import org.zenbot.szolnok.timetable.domain.BusRoute;

import java.util.ArrayList;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.same;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;

public class RemoveBusRoutesExecutionListenerTest {

    private static final String BUS_1 = "BUS_1";
    private static final String BUS_2 = "BUS_2";
    private RemoveBusRoutesExecutionListener testSubject;

    @Mock
    private BusStopRepository busStopRepository;

    @Mock
    private BusRepository busRepository;

    @Mock
    private TimetableResourceProperties properties;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);
        testSubject = new RemoveBusRoutesExecutionListener(busRepository, busStopRepository, properties);
    }

    @Test
    public void beforeJobWithSelectedBusesShouldOnlyRemoveRoutesForThem() {
        // GIVEN
        List<String> selectedBuses = new ArrayList<>();
        selectedBuses.add(BUS_1);
        given(properties.getSelectedBuses()).willReturn(selectedBuses);

        List<Bus> buses = new ArrayList<>();
        Bus bus1 = getBus(BUS_1);
        buses.add(bus1);

        Bus bus2 = getBus(BUS_2);
        buses.add(bus2);

        given(busRepository.findAll()).willReturn(buses);

        // WHEN
        testSubject.beforeJob(new JobExecution(0L));

        //THEN
        verify(busRepository).findAll();
        verify(busRepository).saveAll(same(buses));
        verify(busStopRepository, never()).deleteAll();

        assertThat(bus1.getBusRoutes()).isEmpty();
        assertThat(bus2.getBusRoutes()).isNotEmpty();
    }

    @Test
    public void beforeJObWithNoSelectedBusesShouldRemoveAll() {
        // GIVEN
        given(properties.getSelectedBuses()).willReturn(new ArrayList<>());

        List<Bus> buses = new ArrayList<>();
        Bus bus1 = getBus(BUS_1);
        buses.add(bus1);

        Bus bus2 = getBus(BUS_2);
        buses.add(bus2);

        given(busRepository.findAll()).willReturn(buses);

        // WHEN
        testSubject.beforeJob(new JobExecution(0L));

        //THEN
        verify(busRepository).findAll();
        verify(busRepository).saveAll(same(buses));
        verify(busStopRepository).deleteAll();

        assertThat(bus1.getBusRoutes()).isEmpty();
        assertThat(bus2.getBusRoutes()).isEmpty();
    }

    private Bus getBus(String busName) {
        Bus bus = new Bus();
        bus.setBusName(busName);
        ArrayList<BusRoute> busRoutes = new ArrayList<>();
        busRoutes.add(new BusRoute());
        bus.setBusRoutes(busRoutes);
        return bus;
    }
}

File: szolnok-timetable-backend-batch-bus/src/test/kotlin/org/zenbot/szolnok/timetable/backend/batch/bus/batch/step/bus/listener/RemoveBusRoutesExecutionListenerTest.kt
package org.zenbot.szolnok.timetable.backend.batch.bus.batch.step.bus.listener

import org.junit.Before
import org.junit.Test
import org.mockito.BDDMockito.anyString
import org.mockito.BDDMockito.given
import org.mockito.Mock
import org.mockito.Mockito.any
import org.mockito.Mockito.mock
import org.mockito.Mockito.never
import org.mockito.Mockito.verify
import org.mockito.MockitoAnnotations
import org.springframework.batch.core.JobExecution
import org.springframework.batch.core.JobParameters
import org.zenbot.szolnok.timetable.backend.batch.utils.common.properties.TimetableProperties
import org.zenbot.szolnok.timetable.backend.batch.utils.common.properties.TimetableResourceProperties
import org.zenbot.szolnok.timetable.backend.domain.entity.bus.BusEntity
import org.zenbot.szolnok.timetable.backend.repository.BusRepository

class RemoveBusRoutesExecutionListenerTest {

    private lateinit var testSubject: RemoveBusRoutesExecutionListener

    @Mock
    private lateinit var busRepository: BusRepository

    private lateinit var properties: TimetableProperties

    @Before
    fun setUp() {
        MockitoAnnotations.initMocks(this)

        properties = TimetableProperties()
        properties.resource = TimetableResourceProperties()

        testSubject = RemoveBusRoutesExecutionListener(busRepository)
    }

    @Test
    fun `beforeJob should remove all buses if selected buses is empty`() {
        // GIVEN
        val jobExecution = mock(JobExecution::class.java)
        val jobParameters = mock(JobParameters::class.java)

        given(jobExecution.jobParameters).willReturn(jobParameters)
        given(jobParameters.getString(anyString(), anyString())).willReturn("")

        // WHEN
        testSubject.beforeJob(jobExecution)

        // THEN
        verify(jobExecution).jobParameters
        verify(jobParameters).getString("selectedBuses", "")
        verify(busRepository).deleteAll()
        verify(busRepository, never()).delete(any(BusEntity::class.java))
        verify(busRepository, never()).findAll()
    }

    @Test
    fun `beforeJob should not remove the bus which is not present in the selected buses list`() {
        // GIVEN
        val jobExecution = mock(JobExecution::class.java)
        val jobParameters = mock(JobParameters::class.java)
        val buses = mutableListOf(BusEntity(busName = "2"))

        given(jobExecution.jobParameters).willReturn(jobParameters)
        given(jobParameters.getString(anyString(), anyString())).willReturn("1")
        given(busRepository.findAll()).willReturn(buses)

        // WHEN
        testSubject.beforeJob(jobExecution)

        // THEN
        verify(jobExecution).jobParameters
        verify(jobParameters).getString("selectedBuses", "")
        verify(busRepository).findAll()
        verify(busRepository, never()).deleteAll()
        verify(busRepository, never()).delete(any(BusEntity::class.java))
    }

    @Test
    fun `beforeJob should remove the bus which is present in the selected buses list`() {
        // GIVEN
        val jobExecution = mock(JobExecution::class.java)
        val jobParameters = mock(JobParameters::class.java)
        val bus = BusEntity(busName = "1")
        val buses = mutableListOf(bus)

        given(jobExecution.jobParameters).willReturn(jobParameters)
        given(jobParameters.getString(anyString(), anyString())).willReturn("1")
        given(busRepository.findAll()).willReturn(buses)

        // WHEN
        testSubject.beforeJob(jobExecution)

        // THEN
        verify(jobExecution).jobParameters
        verify(jobParameters).getString("selectedBuses", "")
        verify(busRepository).findAll()
        verify(busRepository).delete(bus)
        verify(busRepository, never()).deleteAll()
    }
}
